apidef/rate_limit_smoothing.go

```diff
@@ -0,0 +1,102 @@
+package apidef
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+// RateLimitSmoothing holds the rate smoothing configuration in effect.
+type RateLimitSmoothing struct {
+	// Enabled if true will enable rate limit smoothing.
+	Enabled bool `json:"enabled" bson:"enabled"`
+
+	// Treshold is the value above which gateway will apply smoothing.
+	Threshold int64 `json:"threshold" bson:"threshold"`
+
+	// Triger holds a value between 0..1 and is used as a percentage value of the
+	// rate limit step to reach, before triggering a SmoothingUp event. Similarly the
+	// value is also used to decrease the allowance when rate limits decrease.
+	//
+	// Example:
+	//
+	// - Allowance 600,
+	// - Current rate 500,
+	// - Step 100,
+	// - Trigger 0.5
+	//
+	// To trigger a RateLimitSmoothingUp event, the rate needs to exceed 550.
+	// The new allowance is calculated as: `Allowance - Trigger * Step`.
+	//
+	// To trigger a RateLimitSmoothingDown event, rate needs to fall below 450.
+	// The new allowance is calculated as: `Allowance - Trigger - (Trigger*Step)`
+	Trigger float64 `json:"trigger" bson:"trigger"`
+
+	// Step defines the step amount for a rate smoothing increase or decrease.
+	Step int64 `json:"step" bson:"step"`
+
+	// Delay is the amount of seconds to wait between allowance updates.
+	Delay int64 `json:"delay" bson:"delay"`
+
+	// Allowance is the current allowance in effect. It's not
```

Comment:
I take it this doesn't appear in the Godocs... but perhaps for complete clarity this is "the currently enforced rate limit" (and is bounded by the `threshold` and `rate`)?

Reply:
It does. We don't need complete clarity, this is a very small scope (5-7 fields over about 40 lines of code).